core/finality-grandpa/src/communication/mod.rs

```diff
@@ -0,0 +1,560 @@
+
+	/// Send a message to a bunch of specific peers, even if they've seen it already.
+	fn send_message(&self, who: Vec<network::PeerId>, data: Vec<u8>);
+
+	/// Inform peers that a block with given hash should be downloaded.
+	fn announce(&self, block: Block::Hash);
+}
+
+/// Create a unique topic for a round and set-id combo.
+pub(crate) fn round_topic<B: BlockT>(round: u64, set_id: u64) -> B::Hash {
+	<<B::Header as HeaderT>::Hashing as HashT>::hash(format!("{}-{}", set_id, round).as_bytes())
+}
+
+/// Create a unique topic for global messages on a set ID.
+pub(crate) fn global_topic<B: BlockT>(set_id: u64) -> B::Hash {
+	<<B::Header as HeaderT>::Hashing as HashT>::hash(format!("{}-GLOBAL", set_id).as_bytes())
+}
+
+impl<B, S> Network<B> for Arc<NetworkService<B, S>> where
+	B: BlockT,
+	S: network::specialization::NetworkSpecialization<B>,
+{
+	type In = NetworkStream;
+
+	fn messages_for(&self, topic: B::Hash) -> Self::In {
+		let (tx, rx) = oneshot::channel();
+		self.with_gossip(move |gossip, _| {
+			let inner_rx = gossip.messages_for(GRANDPA_ENGINE_ID, topic);
+			let _ = tx.send(inner_rx);
+		});
+		NetworkStream { outer: rx, inner: None }
+	}
+
+	fn register_validator(&self, validator: Arc<dyn network_gossip::Validator<B>>) {
+		self.with_gossip(
+			move |gossip, context| gossip.register_validator(context, GRANDPA_ENGINE_ID, validator)
+		)
+	}
+
+	fn gossip_message(&self, topic: B::Hash, data: Vec<u8>, force: bool) {
+		let msg = ConsensusMessage {
+			engine_id: GRANDPA_ENGINE_ID,
+			data,
+		};
+		self.with_gossip(
+			move |gossip, ctx| gossip.multicast(ctx, topic, msg, force)
+		)
+	}
+
+	fn send_message(&self, who: Vec<network::PeerId>, data: Vec<u8>) {
+		let msg = ConsensusMessage {
+			engine_id: GRANDPA_ENGINE_ID,
+			data,
+		};
+
+		self.with_gossip(move |gossip, ctx| for who in &who {
+			gossip.send_message(ctx, who, msg.clone())
+		})
+	}
+
+	fn announce(&self, block: B::Hash) {
+		self.announce_block(block)
+	}
+}
+
+/// A stream used by NetworkBridge in its implementation of Network.
+pub struct NetworkStream {
+	inner: Option<mpsc::UnboundedReceiver<network_gossip::TopicNotification>>,
+	outer: oneshot::Receiver<mpsc::UnboundedReceiver<network_gossip::TopicNotification>>
+}
+
+impl Stream for NetworkStream {
+	type Item = network_gossip::TopicNotification;
+	type Error = ();
+
+	fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+		if let Some(ref mut inner) = self.inner {
+			return inner.poll();
+		}
+		match self.outer.poll() {
+			Ok(futures::Async::Ready(mut inner)) => {
+				let poll_result = inner.poll();
+				self.inner = Some(inner);
+				poll_result
+			},
+			Ok(futures::Async::NotReady) => Ok(futures::Async::NotReady),
+			Err(_) => Err(())
+		}
+	}
+}
+
+/// Bridge between the underlying network service, gossiping consensus messages and Grandpa
+pub(crate) struct NetworkBridge<B: BlockT, N: Network<B>> {
+	service: N,
+	validator: Arc<GossipValidator<B>>,
```

Comment:
If I understand it correctly, there is no reason to share the validator itself with grandpa via the network bridge, instead all the operations below on `self.validator` could be forwarded to the validator held by the network via a message sent by the service. 

For example `global_communication` consists of:

1.`self.validator.note_set`
2. `let incoming = self.service.messages_for`
3. doing something with incoming. 

Here `messages_for` is implemented via messaging to the network thread, so one could also do `validator.note_set` via a message on the same channel, and the ordering would be guaranteed. 

One could for example do something similar to what is done with `register_validator` like:

```rust
self.with_gossip(
	move |gossip, context| {
                let validator = gossip.get_validator(GRANDPA_ENGINE_ID);
                validator.note_set(set_id, &context); // passing context to do the multicast
        }
)
```
Main benefit would be being able to get rid of the `inner: parking_lot::RwLock<Inner<Block>>`, instead just putting the various member currently on `Inner` directly on `GossipValidator`, and only mutating those in responses to messages received by the network/protocol thread.

Reply:
> Main benefit would be being able to get rid of the inner: `parking_lot::RwLock<Inner<Block>>`

I'm not convinced that is a benefit, if the alternative is _more_ costly.

Reply:
@rphmeier @gterzian have either of you run benchmarks?  What are we using for mesage passing?

That said, I agree with @gterzian.  IMO, message passing is easier to debug, and has lower risk of deadlocks.

Reply:
@DemiMarie-parity the communication to network(things like `with_gossip`) is implemented with a crossbeam channel. 

I wonder if benchmarks are required to make this call, and I think we can reason about the problem logically and come to the conclusion that we are unnecessarily creating contention on shared resources(which don't have to be shared). 

And example:

When grandpa does a `round_communication`, it will start with `self.validator.note_round`. `note_round` will lock `inner`, do some setup in the current thread(or rather grandpa task) and then result in sending a message to the network thread(via `with_gossip`, which is what is used as part of `multicast_neighbor_packet`. 

Later, still as part of `round_communication`, grandpa will do a `messages_for`, which again involves sending a message to the network thread. 

Logically, grandpa is communicating with network, and with the validator that lives in network. it just so happens that one part of that communication is implemented via shared state. The shared state does not appear inherent to the logical operations that are performed. 

Another example: `note_set` will lock `inner`, and if network is doing a `propagate` at the same time, you'll get a contention(and if it's grandpa that ends up having to wait for the lock being available, you run into the issue that this can block other tasks running on the same thread). 

Any `propagate` that would happen concurrently to a `note_set` should not be influenced by it, so what we have now is unnecessary contention. With message-passing, the `note_set` will result in a message sent to network/gossip, which will handle it sometimes later, without influencing an ongoing `propagate`(although it will have effect on a `propagate` happening after that message has been handled, which gives us in my opinion an easy to reason about transition of state).

Reply:
Locks make it a lot easier to avoid races  -- I have a feeling we already have some subtle races in the networking code that's been ported to use channels and I'm not too eager to introduce more. For code that's not in a bottlenecking path, that kind of rewrite adds much more code complexity (hours of dev time in design, review, debugging) for a pretty nebulous gain. 

Unless there's an argument other than "locks are bad, avoid at all costs", then I don't see any reason to make that change.

Reply:
> Locks make it a lot easier to avoid races

The re-design I describe above would mean that sate would not be shared among threads anymore, which can play a role in avoided races as well. Weird ordering of messages is in my opinion much easier to debug, one can just add a `println` at the point of message-handling. In my experience it can be harder trying to figure out when something in an `Arc` is unexpectedly mutated. 

> we already have some subtle races in the networking code that's been ported to use channels 

I don't think we should rely on blocks across threads in the implementation of business logic. If removing blocks exposes races, it's an opportunity to improve the situation. 

> Unless there's an argument other than "locks are bad, avoid at all costs"

I think I have provided some above, for this specific case. 

> Parking lot RwLocks are only a single atomic operation in the uncontended case, and I don't see why this lock would be be contended unless under a sustained, multi-peer DoS. 

Ok, so let me provide an example of something I don't think requires Dos for it to result in contention: in [`propagate`](https://github.com/paritytech/substrate/pull/2110/files#diff-d60084fea5b476ff1453a50945ad4d8dR129) a read-lock is acquired and held for the duration of the iteration over messages to be propagated. Re the fast un-contented case(or the efficient spinning a few rounds and trying again) for `parking_lot`, I think you're more likely to hit it when you do things like quick "one operation" acquire/release of the lock, as opposed to making that entire loop a critical section. 

> There are probably 20 other optimizations we've already filed that give 10-100x bigger performance wins.

None of my arguments above relate strongly to performance, rather it's about the design of the system. 

> The general policy in this codebase is to not block PRs that are strictly better than what's currently implemented on improvements that could be made to them.

I don't intend to "block" this PR, I would just like to point out an alternative design. We can discuss further via an issue.